Eigen 3.4 vs Matrix Expression operator==

Sub-task of #15142.

An error message from that PR build:

```
In file included from external/eigen/Eigen/Core:256:0,
                 from external/eigen/Eigen/Dense:1,
                 from drake/common/eigen_types.h:14,
                 from drake/solvers/rotation_constraint.h:8,
                 from solvers/rotation_constraint.cc:1:
external/eigen/Eigen/src/Core/functors/StlFunctors.h: In instantiation of 'bool Eigen::numext::equal_to<T>::operator()(const T&, const T&) const [with T = drake::symbolic::Expression]':
external/eigen/Eigen/src/Core/CoreEvaluators.h:769:22:   required from 'Eigen::internal::binary_evaluator<Eigen::CwiseBinaryOp<BinaryOp, Lhs, Rhs>, Eigen::internal::IndexBased, Eigen::internal::IndexBased>::CoeffReturnType Eigen::internal::binary_evaluator<Eigen::CwiseBinaryOp<BinaryOp, Lhs, Rhs>, Eigen::internal::IndexBased, Eigen::internal::IndexBased>::coeff(Eigen::Index, Eigen::Index) const [with BinaryOp = Eigen::numext::equal_to<drake::symbolic::Expression>; Lhs = const Eigen::Matrix<drake::symbolic::Expression, 4, 4, 0, 4, 4>; Rhs = const Eigen::Transpose<const Eigen::Matrix<drake::symbolic::Expression, 4, 4, 0, 4, 4> >; typename Eigen::internal::traits<typename Eigen::CwiseBinaryOp<BinaryOp, Lhs, Rhs>::Rhs>::Scalar = drake::symbolic::Expression; typename Eigen::internal::traits<typename Eigen::CwiseBinaryOp<BinaryOp, Lhs, Rhs>::Lhs>::Scalar = drake::symbolic::Expression; Eigen::internal::binary_evaluator<Eigen::CwiseBinaryOp<BinaryOp, Lhs, Rhs>, Eigen::internal::IndexBased, Eigen::internal::IndexBased>::CoeffReturnType = bool; Eigen::Index = long int]'
external/eigen/Eigen/src/Core/BooleanRedux.h:93:13:   required from 'bool Eigen::DenseBase<Derived>::all() const [with Derived = Eigen::CwiseBinaryOp<Eigen::numext::equal_to<drake::symbolic::Expression>, const Eigen::Matrix<drake::symbolic::Expression, 4, 4, 0, 4, 4>, const Eigen::Transpose<const Eigen::Matrix<drake::symbolic::Expression, 4, 4, 0, 4, 4> > >]'
external/eigen/Eigen/src/Core/MatrixBase.h:291:36:   required from 'bool Eigen::MatrixBase<Derived>::operator==(const Eigen::MatrixBase<OtherDerived>&) const [with OtherDerived = Eigen::Transpose<const Eigen::Matrix<drake::symbolic::Expression, 4, 4, 0, 4, 4> >; Derived = Eigen::Matrix<drake::symbolic::Expression, 4, 4, 0, 4, 4>]'
drake/solvers/mathematical_program.h:2351:5:   required from 'std::enable_if_t<is_same_v<typename Derived::Scalar, drake::symbolic::Expression>, drake::solvers::Binding<drake::solvers::PositiveSemidefiniteConstraint> > drake::solvers::MathematicalProgram::AddPositiveSemidefiniteConstraint(const Eigen::MatrixBase<Derived>&) [with Derived = Eigen::Matrix<drake::symbolic::Expression, 4, 4, 0, 4, 4>; std::enable_if_t<is_same_v<typename Derived::Scalar, drake::symbolic::Expression>, drake::solvers::Binding<drake::solvers::PositiveSemidefiniteConstraint> > = drake::solvers::Binding<drake::solvers::PositiveSemidefiniteConstraint>]'
solvers/rotation_constraint.cc:120:44:   required from here
external/eigen/Eigen/src/Core/functors/StlFunctors.h:22:19: error: cannot convert 'drake::symbolic::Formula' to 'bool' in return
     return lhs == rhs;
                   ^~~
```

What _should_ be happening is that this overload wins:

https://github.com/RobotLocomotion/drake/blob/dbb6f4edfefee96e20358a12055dd478820e5fe9/common/symbolic_formula.h#L944-L954

... but for some reason with Eigen 3.4, it is not.

--- a/bindings/pydrake/solvers/mathematicalprogram_py.cc
+++ b/bindings/pydrake/solvers/mathematicalprogram_py.cc
@@ -1089,8 +1089,7 @@ void BindMathematicalProgram(py::module m) {
               const Eigen::Ref<const MatrixX<Expression>>& e) {
             return self->AddPositiveSemidefiniteConstraint(e);
           },
-          doc.MathematicalProgram.AddPositiveSemidefiniteConstraint
-              .doc_1args_constEigenMatrixBase)
+          doc.MathematicalProgram.AddPositiveSemidefiniteConstraint.doc_1args_e)
       .def(
           "AddLinearMatrixInequalityConstraint",
           [](MathematicalProgram* self,

--- a/solvers/mathematical_program.h
+++ b/solvers/mathematical_program.h
@@ -2320,7 +2320,6 @@ class MathematicalProgram {
    * Adds a positive semidefinite constraint on a symmetric matrix of symbolic
    * expressions @p e. We create a new symmetric matrix of variables M being
    * positive semidefinite, with the linear equality constraint e == M.
-   * @tparam Derived An Eigen Matrix of symbolic expressions.
    * @param e Imposes constraint "e is positive semidefinite".
    * @pre{1. e is symmetric.
    *      2. e(i, j) is linear for all i, j
@@ -2342,28 +2341,15 @@ class MathematicalProgram {
    * prog.AddPositiveSemidefiniteConstraint(e);
    * @endcode
    */
-  template <typename Derived>
-  typename std::enable_if_t<
-      std::is_same_v<typename Derived::Scalar, symbolic::Expression>,
-      Binding<PositiveSemidefiniteConstraint>>
-  AddPositiveSemidefiniteConstraint(const Eigen::MatrixBase<Derived>& e) {
-    DRAKE_DEMAND(e.rows() == e.cols());
-    DRAKE_ASSERT(e == e.transpose());
-    const int e_rows = Derived::RowsAtCompileTime;
-    MatrixDecisionVariable<e_rows, e_rows> M{};
-    if (e_rows == Eigen::Dynamic) {
-      M = NewSymmetricContinuousVariables(e.rows());
-    } else {
-      M = NewSymmetricContinuousVariables<e_rows>();
-    }
+  Binding<PositiveSemidefiniteConstraint> AddPositiveSemidefiniteConstraint(
+      const Eigen::Ref<const MatrixX<symbolic::Expression>>& e) {
+    // TODO(jwnimmer-tri) Move this whole function definition into the cc file.
+    DRAKE_THROW_UNLESS(e.rows() == e.cols());
+    DRAKE_ASSERT(CheckStructuralEquality(e, e.transpose().eval()));
+    const MatrixXDecisionVariable M = NewSymmetricContinuousVariables(e.rows());
     // Adds the linear equality constraint that M = e.
     AddLinearEqualityConstraint(
-        e - M, Eigen::Matrix<double, e_rows, e_rows>::Zero(e.rows(), e.rows()),
-        true);
-    const int M_flat_size =
-        e_rows == Eigen::Dynamic ? Eigen::Dynamic : e_rows * e_rows;
-    const Eigen::Map<Eigen::Matrix<symbolic::Variable, M_flat_size, 1>> M_flat(
-        &M(0, 0), e.size());
+        e - M, Eigen::MatrixXd::Zero(e.rows(), e.rows()), true);
     return AddPositiveSemidefiniteConstraint(M);
   }
 

